File: 47_Permutation_II/permuteUnique.py
class Solution(object):
    def permuteUnique(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        result = []
        nums.sort()

        def dfs(stack, nums):
            if not nums:
                if stack not in result:
                    result.append(stack)
                return

            for idx, num in enumerate(nums):
                if idx and nums[idx] == nums[idx - 1]: continue
                dfs(stack + [num], nums[:idx] + nums[idx + 1:])

        dfs([], nums)
        return result

sol = Solution()
nums = [1,1,0,0,1,-1,-1,1]
print sol.permuteUnique(nums)

File: 47_Permutation_II/permuteUnique.cpp
#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
    public:
        vector<vector<int> > permuteUnique(vector<int>& nums) {
            sort(nums.begin(), nums.end());
            vector<vector<int> > res;
            dfs(nums, 0, res);
            return res;
        }
    private:
        void dfs(vector<int> nums, int start, vector<vector<int> > &res) {
            if(start >= nums.size()) {
                res.push_back(nums);
                return;
            }

            for(int i = start; i < nums.size(); i++) {
                if(start != i && nums[start] == nums[i]) { continue; }
                swap(nums[start], nums[i]);
                dfs(nums, start + 1, res);
            }
        }
};
